File: 2 - BackEnd/Vanilla/PHP/3 - lang/4 - variables.php
<?php

/* ##########================########## */
/* ######===--- Sin tipado ---===###### */
/* ##########================########## */

/**
 * Variables sin tipado.
 * 
 * Aqui las variables no tienen un tipado definido, sino que pueden adquirir cualquier 
 * valor de manera dinamica.
 */

$variable = 50; // int
$variable = true; // bool.
$variable = 5.555; // float.
$variable = new Usuario(); // Usuario

/* ##########=====================########## */
/* ######===--- Tipado opcional ---===###### */
/* ##########=====================########## */

/**
 * Tambien tenemos tipado en las ultimas versiones de PHP.
 */

(int) $variable = 50;
(bool) $variable = true;
(float) $variable = 5.555;
(object) $variable = new Usuario();

/**
 * NOTA: Apesar de que existe el tipado, (este no es obligatorio).
 */

/* ##########===========########## */
/* ######===--- Scope ---===###### */
/* ##########===========########## */

// -------------------- //
// ------ Global ------ //
// -------------------- //

# Al ser declarada fuera, dentro de este archivo, (es global).
# Accesible para todas las funciones y demas.
$variable_global = 70;

/**
 * Utilizando ($GLOBALS).
 */

function crear_global(){
	// Variable global para todos los archivos PHP.
	$GLOBALS["nombre"] = "valor";
}
crear_global();

// Utilizamos la variable global declarada aun dentro de la funcion.
printf("%s", $GLOBALS["nombre"]);

// --------------------- //
// ------ Locales ------ //
// --------------------- //

# Varible global, pero la funcion no tiene acceso a ella.
$local = true;

function funcion(){
	// Variable (local) solo existente en este bloque de codigo.
	$variable_local = "valor";
}

// ------------------------------------ //
// ------ Estáticas en funciones ------ //
// ------------------------------------ //

function staticExample() {
	// Ámbito estático de función.
    static $staticVariable = 0;

    $staticVariable++;
    
    echo $staticVariable;
}

staticExample(); // Imprime 1
staticExample(); // Imprime 2

// ----------------------- //
// ------ (Closure) ------ //
// ----------------------- //

$outsideVar = 5;

/* La palabra reservada (use) se utiliza cuando se quiere que la funcion anonima acceda a una 
variable global que no esta dentro de su ambito, a esto se le conoce como (Closure). */
$myClosure = function () use ($outsideVar) {
	// Acceso a variable externa.
    echo $outsideVar;
};

$myClosure();

###### ---------------------------------------------------------------------------------------------- ######

// Otra forma de acceder a las variables locales.

$local = ["dato"];

function my_function(){
	// Accede a la variable que se encuentra fuera de ella.
	global $local[0];
}

/* ##########================########## */
/* ######===--- Constantes ---===###### */
/* ##########================########## */

// De esta manera se declaran las constantes.
const constante = 50;
define("constante", 50);

/* ##########=================########## */
/* ######===--- Referencias ---===###### */
/* ##########=================########## */

// --------------------------------- //
// ------ Referencia con ($$) ------ //
// --------------------------------- //

$variable = "Hola mundo.";
$segunda_variable = "Hola amigos";

echo $$variable; # "Hola amigos"

/**
 * Esto sucede porque al poner el doble signo '$$' se hace referencia a la segunda variable.
 */

// -------------------------------- //
// ------ Referencia con (&) ------ //
// -------------------------------- //

/**
 * (&) guarda como referencia una variable en otra.
 */

# Valor inicial para ($var) de 80.
$var = 80;

# Guarda la referencia.
$ref_var = &$var;

# Ahora el valor de ($var) es de 50.
$ref_var = 50;

File: 1 - BackEnd/Vanilla/Python/1 - lang/4 - variables.py
# ##########================##########
# ######===--- Sin tipado ---===######
# ##########================##########

# Variables sin tipado.
variable = 50 # int
variable = True # bool
variable = 5.555 # float

# Instancia de objeto.
variable = Usuario()  # Usuario (suponiendo que Usuario es una clase definida)

# ##########============##########
# ######===--- Tipado ---===###### 
# ##########============##########

# Python no requiere la declaración de tipo explícita.
# Las variables se definen y el tipo se infiere automáticamente.

variable = 50        # int
variable = True      # bool
variable = 5.555    # float
variable = Usuario()  # object (suponiendo que Usuario es una clase definida)

# ##########===========##########
# ######===--- Scope ---===######
# ##########===========##########

# -------------------- #
# ------ Global ------ #
# -------------------- #

'''
    Acceder a las variables externas.
'''

# Otra forma de acceder a las variables locales.
local = ["dato"]

def my_function():
    # Accede a la variable externa usando `global` solo para variables globales, no para listas.
    global local
    print(local[0])

'''
    Modificar variables externas.
'''

# Variable global.
nombre = 70

def crear_global():
    global nombre
    # Variable global accesible en todas partes.
    nombre = "valor"

crear_global()
print(nombre)  # Imprime "valor"

# --------------------- #
# ------ Locales ------ #
# --------------------- #

# Variable global, pero la función NO tiene acceso a ella.
local = True

def funcion():
    # Variable local.
    variable_local = "valor"

# ------------------------------------ #
# ------ Estáticas en funciones ------ #
# ------------------------------------ #

# Ámbito estático de función.

def myFuncion():
    # (__dict__) es el diccionario de atributos de la funcion.
    if 'variable' not in myFuncion.__dict__:
        # Si no existe la inicializa.
        myFuncion.variable = 0

    myFuncion.variable += 1

    print(myFuncion.variable)

myFuncion()  # Imprime 1
myFuncion()  # Imprime 2

# ----------------------- #
# ------ (Closure) ------ #
# ----------------------- #

variable = 5

# Función anónima (closure) accede a la variable externa.
my_closure = lambda: print(variable)

my_closure()  # Imprime 5

# ##########================##########
# ######===--- Constantes ---===###### 
# ##########================##########

# Definir constantes en Python no es estricto, (pero puedes usar la convencion).
CONSTANTE = 50

# ##########=================##########
# ######===--- Referencias ---===###### 
# ##########=================##########

# ----------------- #
# ------ (&) ------ #
# ----------------- #

# En Python, se puede trabajar con referencias de manera indirecta.